### ISSUE TITLE
Pyright should not report "TypeVar appears only once in generic function" for TypeVar in `self`

### ISSUE THREAD

**Issue by yaxum62** [microsoft/pyright#6798]
With `reportInvalidTypeVarUse`, current version of Pyright reports error for following code.

```python
T = TypeVarTuple("T")
X = TypeVar("X")
Y = TypeVar("Y")

class Foo(Generic[*T]):
  @overload
  def fn(self: 'Foo[X]'):
      ...
  @overload
  def fn(self: 'Foo[X, Y]'):
      ...
```
with error message `TypeVar "X"/"Y" appears only once in generic function signature`. However this use of TypeVar should be valid since it applies to 'self', used as restrictions instead of variation of parameter type. A similar usage is included in https://peps.python.org/pep-0646/#overloads-for-accessing-individual-types. The example in pep just happened to use all TypeVars in return type so this error won't trigger but that is not always the case.

Ideally there should be a wildcard TypeVar for this, like:
```python
class Foo(Generic[*T]):
  @overload
  def fn(self: 'Foo[_]'):
    ...
  @overload
  def fn(self: 'Foo[_, _]'):
    ...
```
But in current state without wildcard, pyright should at least spare the ugly but valid use of TypeVar.  

pylance vscode extension version `v2023.12.1`

**Comment by erictraut** [comment]
This behavior is intentional, not a bug. The TypeVar in this case serves no purpose. You could simply replace it with `Any`, and it would have the same meaning.

**Comment by yaxum62** [comment]
Ok, `Any` could work as a wild card, what about `TypeVarTuple` in self? e.g.

```python
T = TypeVarTuple("T")
Tail = TypeVarTuple("Tail")

class Foo(Generic[*T]):
    @overload
    def fn(self: 'Foo[()]'):
        ...
    @overload
    def fn(self: 'Foo[Any, *Tail]'):
        ...
```
pyright reports `TypeVar "Tail" appears only once in generic function signature`

**Comment by yaxum62** [comment]
Never mind, just realized `*Any` is a thing.

**Comment by erictraut** [comment]
I'd recommend `Foo[Any, *tuple[Any, ...]]` rather than `Foo[Any, *Any]`. The former is more conformant with PEP 646 and the typing spec. I'm kind of surprised that the latter works, and it may stop working at some point.
